File: packages/sanbox/src/bubblewrap.ts
import type { SandboxPolicy } from './types';
import runtime from './runtime-policy.json';

export function bubblewrapArgs(policy: SandboxPolicy, cwd: string): string[] {
    const args = [...runtime.linux.args];
    if (policy.network !== 'allow') args.push('--unshare-net');
    for (const path of runtime.linux.runtimePaths) args.push('--ro-bind-try', path, path);
    args.push('--proc', '/proc', '--dev', '/dev', '--tmpfs', '/tmp');
    if (policy.network === 'allow') {
        for (const path of runtime.linux.networkPaths) args.push('--ro-bind-try', path, path);
    }
    const grants = [
        ...new Set(policy.readOnlyPaths ?? []), ...new Set(policy.writablePaths ?? []),
    ].sort((a, b) => a.length - b.length);
    for (const path of grants) args.push(policy.writablePaths?.includes(path) ? '--bind' : '--ro-bind', path, path);
    args.push('--chdir', cwd, '--');
    return args;
}

File: packages/sanbox/native/src/bubblewrap.rs
use std::process::Command;
use crate::{LinuxPolicy, Mount, NetworkAccess};

pub(crate) fn command(script: &str, cwd: &str, mounts: &[Mount], network: NetworkAccess, policy: &LinuxPolicy) -> Command {
    let mut result = Command::new("/usr/bin/bwrap");
    result.args(&policy.args);
    if network == NetworkAccess::Deny { result.arg("--unshare-net"); }
    for path in &policy.runtime_paths { result.args(["--ro-bind-try", path, path]); }
    result.args(["--proc", "/proc", "--dev", "/dev", "--tmpfs", "/tmp"]);
    if network == NetworkAccess::Allow {
        for path in &policy.network_paths { result.args(["--ro-bind-try", path, path]); }
    }
    for mount in mounts {
        result.args([if mount.writable { "--bind" } else { "--ro-bind" }, &mount.source, &mount.target]);
    }
    result.args(["--chdir", cwd, "--", "bash", "--noprofile", "--norc", "-c", script]);
    result
}
